loader/loader_windows.c

```diff
@@ -793,6 +793,133 @@ VkResult windows_read_data_files_in_registry(const struct loader_instance *inst,
+        sorted_array[*sorted_devices_count].physical_devices = NULL;
+        // Unless OOHM occurs, only return VK_SUCCESS
+        if (res != VK_ERROR_OUT_OF_HOST_MEMORY) {
+            res = VK_SUCCESS;
+            loader_log(inst, VULKAN_LOADER_WARN_BIT, 0, "Failed to convert DXGI adapter into Vulkan physical device");
+        }
+        return res;
+    }
+
+    // Because the loader calls EnumerateAdapterPhysicalDevices on all drivers with each DXGI Adapter, if there are multiple drivers
+    // that share a luid the physical device will get queried multiple times. We can prevent that by not adding them if the
+    // enumerated physical devices have already been added.
+    bool already_enumerated = false;
+    for (uint32_t j = 0; j < *sorted_devices_count; j++) {
+        if (count == sorted_array[j].device_count) {
+            bool matches = true;
+            for (uint32_t k = 0; k < sorted_array[j].device_count; k++) {
+                if (sorted_array[j].physical_devices[k] != sorted_array[*sorted_devices_count].physical_devices[k]) {
+                    matches = false;
+
+                    break;
+                }
+            }
+            if (matches) {
+                already_enumerated = true;
+            }
+        }
+    }
+    if (!already_enumerated) {
+        sorted_array[*sorted_devices_count].device_count = count;
+        sorted_array[*sorted_devices_count].icd_index = icd_idx;
+        sorted_array[*sorted_devices_count].icd_term = icd_term;
+        sorted_array[*sorted_devices_count].windows_adapter_luid = luid;
+        (*sorted_devices_count)++;
+    }
+
+    return VK_SUCCESS;
+}
+
+// Whenever there are multiple drivers for the same hardware and one of the drivers is an implementation layered ontop of another
+// API (such as the Dozen driver which converts vulkan to Dx12), we want to make sure the layered driver appears after the 'native'
+// driver. This function iterates over all physical devices and make sure any with matching LUID's are sorted such that drivers with
+// a underlyingAPI of VK_LAYERED_DRIVER_UNDERLYING_API_D3D12_MSFT are ordered after drivers without it.
+void sort_physical_devices_with_same_luid(struct loader_instance *inst, uint32_t sorted_devices_count,
+                                          struct loader_phys_dev_per_icd *sorted_array) {
+    bool app_is_vulkan_1_1 = loader_check_version_meets_required(LOADER_VERSION_1_1_0, inst->app_api_version);
+
+    for (uint32_t i = 0; sorted_devices_count > 1 && i < sorted_devices_count - 1; i++) {
+        for (uint32_t j = i + 1; j < sorted_devices_count; j++) {
+            // Only want to reorder physical devices if their ICD's LUID's match
+            if ((sorted_array[i].windows_adapter_luid.HighPart != sorted_array[j].windows_adapter_luid.HighPart) &&
+                (sorted_array[i].windows_adapter_luid.LowPart != sorted_array[j].windows_adapter_luid.LowPart)) {
```

Comment:
```suggestion
            if ((sorted_array[i].windows_adapter_luid.HighPart != sorted_array[j].windows_adapter_luid.HighPart) ||
                (sorted_array[i].windows_adapter_luid.LowPart != sorted_array[j].windows_adapter_luid.LowPart)) {
```

Or

```suggestion
            if (!((sorted_array[i].windows_adapter_luid.HighPart == sorted_array[j].windows_adapter_luid.HighPart) &&
                (sorted_array[i].windows_adapter_luid.LowPart == sorted_array[j].windows_adapter_luid.LowPart))) {
```

The high bits of a LUID are often 0 and this will result in swapping adapters that only have matching high bits.

Reply:
I forgot De Morgan's laws! Very grateful for catching this.